pytorch_lightning/core/lightning.py

```diff
@@ -1326,6 +1326,23 @@ def training_step(...):
         self.trainer.train_loop.backward(loss, optimizer=None, opt_idx=None, *args, **kwargs)
         self._running_manual_backward = False
 
+    def on_before_backward(self, loss: torch.Tensor) -> None:
+        """
+        Override on_before_backward with your own implementation if you need to.
```

Comment:
- it might not be clear to users that this is only called with automatic optimization. that also can lead to confusion over when to override this, optimizer step, or manual backward  and  when to set automatic_optimization=False.
- what's the guidance on this defined in the lightning module vs in core/hooks.py under ModelHooks?

i feel like if we could easily offer wrapped optimizers to users, would we still need all these intermediary hooks? would we feel more comfortable encouraging users to do manual optimization?

Reply:
Hey @ananthsub,

For 1. This is being called independently of `automatic optimization`.
See below:
<img width="495" alt="Screenshot 2021-06-08 at 11 05 44" src="https://user-images.githubusercontent.com/12861981/121167031-7a013780-c849-11eb-9766-15ff0e3e7241.png">

For 2. For consistency reason, as ``on_after_backward`` is both a LightningModule / Callback hooks, it should be the same with ``on_before_backward``.

Not sure what you mean with `wrapped optimizers to users`. To me, this won't follow PyTorch API and I still prefer hooks through Callbacks and LightningModule.

``manual optimization`` is more about controlling the flow around ``backward`` and ``optimizer`` call. The hooks should still be called properly.

Adding hooks shoudn't make the ``manual optimization`` experience lower.

Best,
T.C